### ISSUE TITLE
Dataset of Mega Man Maker levels

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#171]
If we want to train a diffusion model, then we're probably going to need thousands of Mega Man Maker levels, let's say 10,000. So, what is an easy way to get 10,000 levels? We certainly can't download each one manually, and we can't even realistically look at each one. See if you can automate a script to download a bunch of levels. Maybe we just start at level ID 1 and match through the numbers to see how many we can download.

Whatever you download, don't commit the files to the repo. We're in very initial testing phase here. This is just the starting point. Report on your findings in the issue thread as you go.

**Comment by almzayena-lab** [comment]
Okay so far I made a file called bulk with bulk_download in it... I ran it to save id's 1-20 to check if it works in general. 3 levels out of the 20 successfully downloaded thus far.

<img width="923" height="534" alt="Image" src="https://github.com/user-attachments/assets/2835c952-3e22-47ca-ae5d-0903f3a7fa44" />
<img width="937" height="480" alt="Image" src="https://github.com/user-attachments/assets/784c6c38-a9e0-40aa-8af5-f22cea3396d2" />

**Comment by schrum2** [comment]
We shouldn't be surprised if there are too many gaps, and the lowest numbers (presuambly the earliest contributions) might have lower quality. Being able to download any levels in a range is useful though.

I wonder if there is a way to access metrics about the levels such as their ratings, number of plays, etc. This is definitely online, and there is probably a way to scrape it and use the data to decide what to incorporate into a training set.

**Comment by almzayena-lab** [comment]
Okay so for my code now, I have created a way to only download levels that have more likes than dislikes and have minimum of 1 download. I have tested downloading with internals starting from 1-200 which had 29 valid downloads; 10000-10200 which had 35 downloads; 100000-100200 which had 37 downloads. this showed me that the higher we start the more valid levels we can download so i am going to start downloading at 100,000. 

<img width="893" height="586" alt="Image" src="https://github.com/user-attachments/assets/ab56d180-0c8a-492b-9cc0-b467a55e2007" />

**Comment by almzayena-lab** [comment]
<img width="892" height="610" alt="Image" src="https://github.com/user-attachments/assets/ad4c977e-f68b-4a1c-9fd3-18f0be2a19e2" />

I got amazing results with 67 valid downloads starting from ID 200,000 so i think thats what im going to do.

**Comment by schrum2** [comment]
Add meta data about the levels to the download log, and also give an option for the user to specify how many levels to download ... the code will keep looping through ID numbers until enough levels satisfying the criteria are found.

**Comment by almzayena-lab** [comment]
<img width="932" height="636" alt="Image" src="https://github.com/user-attachments/assets/bd6690a4-b82a-45ea-8293-16647797a9ea" />
<img width="933" height="817" alt="Image" src="https://github.com/user-attachments/assets/e79d5d41-f7ea-433e-97ba-b8bac3281f4a" />
<img width="672" height="973" alt="Image" src="https://github.com/user-attachments/assets/641ec726-fdfb-4fea-821e-56207e2175d2" />

I fixed it so now you can choose how many levels to download with `--target`, and it just keeps looping through IDs starting at 200000 until it gets enough valid ones instead of using a fixed range. I also made the log save in the same MegaMaker Levels folder and only append successful downloads so it doesn’t reset, plus I added metadata like name, author, downloads, likes, and dislikes to each logged level, while the terminal still shows failed/skipped levels with reasons so I can actually see what’s going on while it runs.

**Comment by schrum2** [comment]
Next step of the dataset creation pipeline is to mass convert each MMLV level to VGLC format (the script should have a command line parameter that creates a directory inside the MarioDiffusion repo to store the files, but don't commit any of these to the repo ... maybe even choose to ignore them via .gitignore), and then process all of those levels to create a json for training using the normal VGLC to json steps already described in MM_README. I anticipate some problems ... report on them as they come up

**Comment by almzayena-lab** [comment]
So far i have created a conversion but i am at an error which i will work on on Monday because i will be leaving early. 

<img width="896" height="857" alt="Image" src="https://github.com/user-attachments/assets/34e27dc3-cff4-4a2c-ba10-4f15bc4587af" />

**Comment by schrum2** [comment]
Try loading the offending level in Mega Man Maker and see if it has any obvious problems. Assuming it is playable, that might mean MMM creates a default spawn location when one is missing. We can do the same. @fwilson12 actually made some code for setting start and end points.

**Comment by almzayena-lab** [comment]
<img width="1276" height="710" alt="Image" src="https://github.com/user-attachments/assets/899f426b-1f32-4368-a546-681af06eb5d0" />
<img width="1513" height="1079" alt="Image" src="https://github.com/user-attachments/assets/ae04f332-0c11-47e4-bef7-7783a51fea9b" />

Everything in this step is now done. I created a bulk converter script with an --output parameter, converted all 16 .mmlv levels to VGLC format (excluded from the repo via .gitignore), and ran them through the existing JSON pipeline. For the missing spawn issue, I added a fallback default location in find_start() and wrapped the main loop in a try/except to skip levels with unresolvable branching paths. 10 out of 16 levels parsed successfully.

**Comment by schrum2** [comment]
Just following up on our conversation. What we want is a dataset of about 10,000 samples. That will not require 10,000 levels though, since each level is sliced up into lots of training scenes. Still, make an automated process that selects MMLV levels from online (however many it takes) to get 10,000 training samples to work with. Once you have that, train on it to see how it comes out.

**Comment by almzayena-lab** [comment]
<img width="1919" height="1079" alt="Image" src="https://github.com/user-attachments/assets/cdea98ec-9253-4ea2-bdae-440113babe9f" />

**Comment by almzayena-lab** [comment]
Dataset pipeline is fully working! Successfully converted Mega Man Maker community levels to VGLC format and generated 153,788 training samples using a sliding window approach. Currently training the MLM (text encoder) — 300 epochs. Diffusion model training comes after. Will post results when generation is ready.

**Comment by schrum2** [comment]
I wonder if that is too many samples ... we'll see how slow the training goes, but let it continue for now.

**Comment by schrum2** [comment]
Reiterating some things we discovered from looking at the data:
1. Make sure that the data creation process ignores parts of the MMLV that contain nothing but blank space. These levels are more complex, so the default MMLV to VGLC approach may not work as well
2. Look at the data in the ASCII data browser before training on it
3. Try to restrict the number of training samples to about 10,000. Any more than that will take too long to train on.
4. Save the data to a file with the .json extension. Don't commit it to the repo.
5. Split the data into train/test/validate before training
6. Train the model

**Comment by almzayena-lab** [comment]
<img width="1919" height="1079" alt="Image" src="https://github.com/user-attachments/assets/902fd835-4d41-4e21-aee0-a04b341d38b8" />

I have adjusted the code for mmlv to vglc conversion and the levels being generated are now highly more recognizable and have more known characters. I will now attempt to trains on this data to get some good data.

**Comment by almzayena-lab** [comment]
<img width="903" height="649" alt="Image" src="https://github.com/user-attachments/assets/a7570225-cd12-4f60-9920-0b47019894dd" />

I have uploaded levels and also set a cap for 7,000 samples so that the computer isin't overwhelmed by the number of samples it needs to process.

**Comment by almzayena-lab** [comment]
More clearly these are the commands im using to prepare and train my data. 

python MM_create_ascii_captions.py --dataset datasets/MM_data/json_mmlv --tileset datasets/MM.json --output datasets/MM_data/json_mmlv_captions
python tokenizer.py save --json datasets/MM_data/json_mmlv_captions --pkl_file datasets/MM_data/MM_Tokenizer_mmlv.pkl
python train_mlm.py --epochs 300 --save_checkpoints --json datasets/MM_data/json_mmlv_captions --pkl datasets/MM_data/MM_Tokenizer_mmlv.pkl --output_dir MM-MLM-mmlv --seed 0

**Comment by schrum2** [comment]
Reminder to save data to .json files.

We just removed the code that caps and shuffles the data samples. For now, try training on data from just one level (should not have too much more than 10,000 samples). Post results on how well that looks, and investigate that level a bit to see why it is so large.

If that works fine, then a better way to cut down on the samples would be to increase the window stride. In other words, we still collect data by sliding a window over the data, but instead of one tile at a time we make bigger jumps (pretty sure that code to do this is in the Mario version of the data creation script). However, we have to be careful here because of all the code for navigating corners in Mega Man ... we don't want to mess that up. So, instead of actually increasing the stride, we'll slide through the scenes at the normal rate, but have some sort of variable ensures we only save a sample when the current count modulus the variable equals 0 (so that we can choose to get every 5th, 6th, or nth sample)

**Comment by almzayena-lab** [comment]
<img width="1079" height="858" alt="Image" src="https://github.com/user-attachments/assets/65e6dc74-7141-461d-b9a8-25711c8e1649" />

Okay i was able to get this level and create 8,000 samples from it, however, it seems like the path its travels in the ascii is a bit janky and it doesn't seem to move up and down like its supposed to. Im not sure if its just how its supposed to be or if theres an actual issue but i will try  to figure that out before training anything on it.

**Comment by schrum2** [comment]
Yes, the A* is not perfect. When creatin the datasets, we can choose to filter based on A*, or completely ignore A*. For now, I think ignoring A* is appropriate, though we will fix it a bit more in another issue.

**Comment by schrum2** [comment]
Just a reminder of what is needed to complete this issue:
1. Download a bunch of MMLV levels
2. Convert them to VGLC ASCII
3. Process those into captioned json samples for training
4. Train corresponding MLM and diffusion models to verify

The last post said you had a set of 8,000 samples, which is good, but inspect some of them to make sure they are quality data. We don't want to waste a lot of time training on bad level samples that contain empty void or are obstructed by an impassable wall. I think some of this filtering was already done, but check before training a model. 

In the final message to this thread before closing, post the sequence of instructions used to create and train on the dataset (or the name(s) of the batch file(s) that do it)

**Comment by schrum2** [comment]
Thanks to the instructions in #191 it is clear that this issue is complete, so I will close it. There are still some related follow-up issues, but those can be separate. I'm going to close the issue with some examples of what I trained. 

First, it is clear that the model has good recall of specific captions it saw during training, as shown by these examples comparing original data in the ASCII browser along with generated samples using the same caption in the interactive GUI:

<img width="1586" height="1322" alt="Image" src="https://github.com/user-attachments/assets/656143f8-c960-4b61-8a7a-4596784b7182" />

<img width="1701" height="1324" alt="Image" src="https://github.com/user-attachments/assets/581becdf-5714-4084-be23-b4e142113a09" />

There are plenty of bad examples too though:

<img width="1741" height="1341" alt="Image" src="https://github.com/user-attachments/assets/939c6822-725b-4fed-b243-4a4dd1519418" />

<img width="1782" height="1326" alt="Image" src="https://github.com/user-attachments/assets/1673f2c6-4cdb-45db-a5c5-6efcc8a66dc9" />

Note that even though the first of those does not look like the original sample, the result does have high caption adherence. In contrast, the second one has low scores and doesn't look like the source. That's why there is an issue to add analysis via caption adherence so that we can get numeric results on this. #191 

Another issue #194 notes that there are still problems with the training data. For example, this is a bad scene with an odd caption, though the model generats scenes that get the same bad caption:

<img width="1632" height="1323" alt="Image" src="https://github.com/user-attachments/assets/f1f5f97a-4264-402b-908a-7272afd81b99" />

However, if the model is doing a good job generating bad level scenes, then that is a problem for us. It probably explains the mixed quality in these results from run_diffusion:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/b6f3e8bd-805e-43ae-b49e-814f0d930a9f" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/573d4d9f-694b-4c36-b1e1-24361c5b3207" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/8c9bc2d9-3411-4a65-8f37-29f1218e9fa3" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/702f6629-35e6-4738-962e-514e77129e07" />

In any case, I'm closing the issue, but let's keep trying to both improve the training data, and improve oru analysis of the results, as described in the other issues.
